src/librustc_traits/lowering.rs

```diff
@@ -363,10 +365,60 @@ pub fn program_clauses_for_type_def<'a, 'tcx>(
 }
 
 pub fn program_clauses_for_associated_type_def<'a, 'tcx>(
-    _tcx: TyCtxt<'a, 'tcx, 'tcx>,
-    _item_id: DefId,
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    item_id: DefId,
 ) -> Clauses<'tcx> {
-    unimplemented!()
+    // Rule ProjectionEq-Normalize
+    //
+    // ProjectionEq can succeed by normalizing:
+    // ```
+    // forall<Self, P1..Pn, Pn+1..Pm, U> {
+    //   ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U) :-
+    //       Normalize(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> -> U)
+    // }
+    // ```
+    let item = tcx.associated_item(item_id);
+    debug_assert_eq!(item.kind, ty::AssociatedKind::Type);
+    let trait_id = match item.container {
+        ty::AssociatedItemContainer::TraitContainer(trait_ref) => trait_ref,
+        _ => bug!("not an trait container"),
+    };
+
+    // `Self as Trait<P1..Pn>`
+    let trait_ref = tcx.impl_trait_ref(trait_id).unwrap();
+
+    // `U`
+    let ty = tcx.type_of(item_id);
+
+    // `<Self as Trait<A1..An>>::AssocType<Pn+1..Pm>`
+    let projection_ty = ty::ProjectionTy::from_ref_and_name(tcx, trait_ref, item.ident);
+
+    //  ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U)
+    let projectioneq_goal = DomainGoal::Holds(
+        <ty::ProjectionPredicate as Lower<self::WhereClause>>::lower(
+            &ty::ProjectionPredicate {
+                projection_ty,
+                ty,
+            }
+        )
+    );
+
+    // `Normalize(<A0 as Trait<A1..An>>::AssocType<Pn+1..Pm> -> U)`
+    let hypotheses = vec![ty::Binder::dummy(
+        ty::ProjectionPredicate { projection_ty, ty }.lower(),
+    )];
+
+    //  ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U) :-
+    //      Normalize(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> -> U)
+    let clause = ProgramClause {
+        goal: projectioneq_goal,
+        hypotheses: tcx.mk_goals(
+            hypotheses
+                .into_iter()
+                .map(|wc| Goal::from_poly_domain_goal(wc, tcx)),
```

Comment:
As for the above comment, the whole `map` can be dropped leaving just `hypotheses.into_iter()`.